src/recipe_dictionary.cpp

```diff
@@ -0,0 +1,63 @@
+#include "recipe_dictionary.h"
+#include <algorithm> //std::remove
+
+void recipe_dictionary::add( recipe *rec )
+{
+    recipes.push_back( rec );
+    add_to_component_lookup( rec );
+    by_name[rec->ident] = rec;
+    by_index[rec->id] = rec;
+    by_category[rec->cat].push_back( rec );
+}
+
+void recipe_dictionary::remove( recipe *rec ) {
+    recipes.remove( rec );
+    remove_from_component_lookup( rec );
+    by_name.erase( rec->ident );
+    by_index.erase( rec->id );
+    by_category.erase( rec->cat );
+}
+
+void recipe_dictionary::add_to_component_lookup( recipe* r )
+{
+    std::unordered_set<itype_id> counted;
+    for( const auto &comp_choices : r->requirements.components ) {
+        for( const item_comp &comp : comp_choices ) {
+            if( counted.count(comp.type) ) {
+                continue;
+            }
+            counted.insert( comp.type );
+            by_component[comp.type].push_back( r );
+        }
+    }
+}
+
+void recipe_dictionary::remove_from_component_lookup( recipe* r )
+{
+    for( auto &map_item : by_component ) {
+        std::vector<recipe *> &rlist = map_item.second;
+        rlist.erase( std::remove( rlist.begin(), rlist.end(), r ), rlist.end() );
+    }
+}
+
+void recipe_dictionary::clear()
+{
+    by_component.clear();
+    by_name.clear();
+    by_index.clear();
+    by_category.clear();
+    for( auto &recipe : recipes ) {
+        delete recipe;
+    }
+    recipes.clear();
+}
+
+const std::vector<recipe *>& recipe_dictionary::in_category( const std::string &cat )
+{
+    return by_category[cat];
+}
+
+const std::vector<recipe *>& recipe_dictionary::of_component( const itype_id &id )
+{
+    return by_component[id];
+}
```

Comment:
That has to be it, both files I have end with newlines and TortoiseGit is not showing me any changed files.